app/controllers/auth/registrations_controller.rb

```diff
@@ -93,6 +94,15 @@ def invite_code
 
   private
 
+  def check_captcha
+    if ENV['HCAPTCHA_ENABLED'] == 'true' && !verify_hcaptcha
+      self.resource = resource_class.new sign_up_params
+      resource.validate
+	  flash[:alert] = Hcaptcha::Helpers.to_error_message(:verification_failed)
```

Comment:
The indentation isn't right. Also I wonder if there could be a cleaner way to do that, adding to the form's validation errors instead…

Reply:
Oops, thought I double checked indentation.

As to adding the error to the form’s validation errors, can you point me towards a way for me to do that? Sorry if that’s pretty basic, I have been looking but I haven’t found it yet

Reply:
According to https://github.com/firstmoversadvantage/hcaptcha#verify_recaptcha, you can pass the model to `verify_hcaptcha`. Unfortunately, the model is built within the `create` method, so doing that in a `before_action` hook isn't really possible… also it makes me notice the resource isn't initialized quite correctly here.

I think maybe the best way would be to it this way, in the `!verify_hcaptcha` case (untested):
```ruby
      build_resource(sign_up_params)
      resource.validate
      resource.errors.add(:base, flash.delete(:hcaptcha_error))
      respond_with resource
```

Reply:
will look into this, thank you

Reply:
The code you proposed works (pushed in the next commit), and it adds the captcha error to the error count, but doesn’t display the hcaptcha error anywhere:

Examples:
![image](https://user-images.githubusercontent.com/6833300/81403260-4b3efc00-9133-11ea-9e6f-261642c6b0a9.png)
![image](https://user-images.githubusercontent.com/6833300/81403351-7a556d80-9133-11ea-842a-25a0782dcb12.png)

Ideally it would be nice to display "Captcha verification failed" somewhere, perhaps below the captcha itself in the same style as the forms, but I don’t really have much of an idea yet as to how to do that

Reply:
hmmm… that whole thing is a bit annoying because `verify_hcaptcha` code either adds to an initialized model's errors, or adds to a special `flash` message. The issue with the former approach is that ideally you'd want that to check that after the resource is built and checked in the `create` handler (from devise), but we preferably don't want to rewrite the `create` method.

The latter is maybe a bit less satisfying, but it's easier to deal with.

The verification block would become:
```ruby
    if ENV['HCAPTCHA_ENABLED'] == 'true' && !verify_hcaptcha
      build_resource(sign_up_params)
      resource.validate
      render :new
    end
```

`hcaptcha_error` would have to be added to `FlashesHelper.user_facing_flashes` and some CSS would need to be added to handle that case.